raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+
+        // This also allows for registered MSs to deposit and use other MSs
+        balances[beneficiary] += amount;
+
+        // Transfer the deposit to the smart contract
+        require(token.transferFrom(msg.sender, address(this), amount));
+
+        emit NewDeposit(msg.sender, beneficiary, amount);
+    }
+
+    /// @notice Deposit and register as Monitoring Service in one function call
+    /// Can be called once and is only callable by addresses not already registered as MS
+    function depositAndRegisterMonitoringService() isNotMonitor(msg.sender) public {
+        uint amount_to_deposit = minimum_deposit - balances[msg.sender];
+
+        // Deposit the amount needed to register
+        deposit(msg.sender, amount_to_deposit);
+        // Register as a Monitoring Service
+        require(registerMonitoringService());
+    }
+
+    function updateReward(
+        address token_network_address,
+        address closing_participant,
+        address non_closing_participant,
+        bytes reward_proof_signature,
+        uint256 nonce,
+        address reward_sender_address
+    )
+    internal
+    {
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+        bytes32 channel_identifier = token_network.getChannelIdentifier(closing_participant, non_closing_participant);
+        // Get the Reward struct for the correct channel
+        Reward storage reward = rewards[channel_identifier];
+
+        // Only allow PBs with higher nonce to be submitted
+        require(reward.nonce < nonce);
+
+        // MSC stores channel_identifier, MS_address, reward_amount, nonce
+        // of the MS that provided the balance_proof with highest nonce
+        rewards[channel_identifier] = Reward({
+            reward_proof_signature: reward_proof_signature,
+            nonce: nonce,
+            reward_sender_address: reward_sender_address
+        });
+	}
+
+    /// @notice Called by a registered MS, when providing a new balance proof
+    /// to a monitored channel.
+    /// Can be called multiple times by different registered MSs as long as the PB provided
+    /// is newer than the current newest registered BP.
+    /// @param nonce Strictly monotonic value used to order BPs
+    /// omitting PB specific params, since these will not be provided in the future
+    /// @param reward_proof_signature Signature of the Raiden Node signing the reward
+    /// @param token_network_address Address of the Token Network in which the channel
+    /// being monitored exists.
+    function monitor(
+        bytes32 balance_hash,
+        uint256 nonce,
+        bytes32 additional_hash,
+        bytes closing_signature,
+        bytes non_closing_signature,
+        address reward_sender_address,
+        bytes reward_proof_signature,
+        address token_network_address,
+        address closing_participant,
+        address non_closing_participant
+    )
+        isMonitor(msg.sender)
+        public
+    {
+        updateReward(
```

Comment:
Reward proof checks should be done here + checks that the signature belongs to the `non_closing_participant`.